File: java/src/main/java/com/github/apstndb/spanvalue/ValueFormat.java
package com.github.apstndb.spanvalue;

import static com.github.apstndb.spanvalue.ProtoAccess.arrayElementType;
import static com.github.apstndb.spanvalue.ProtoAccess.boolValue;
import static com.github.apstndb.spanvalue.ProtoAccess.fieldName;
import static com.github.apstndb.spanvalue.ProtoAccess.fieldType;
import static com.github.apstndb.spanvalue.ProtoAccess.isNullValue;
import static com.github.apstndb.spanvalue.ProtoAccess.listValues;
import static com.github.apstndb.spanvalue.ProtoAccess.numberValue;
import static com.github.apstndb.spanvalue.ProtoAccess.protoTypeFqn;
import static com.github.apstndb.spanvalue.ProtoAccess.stringValue;
import static com.github.apstndb.spanvalue.ProtoAccess.structFields;
import static com.github.apstndb.spanvalue.ProtoAccess.typeCode;
import static com.github.apstndb.spanvalue.ProtoAccess.valueKind;
import static com.github.apstndb.spanvalue.TypeFormat.formatTypeCode;
import static com.github.apstndb.spanvalue.TypeFormat.formatTypeVerbose;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

/** FormatConfig presets and value formatting. */
public final class ValueFormat {
  private ValueFormat() {}

  public enum Preset {
    SIMPLE,
    LITERAL,
    SPANNER_CLI
  }

  public record FormatConfig(
      Preset preset, String nullString, Quote.LiteralQuoteConfig quote) {

    public FormatConfig {
      if (nullString == null || nullString.isEmpty()) {
        throw new EmptyNullStringException("null_string must not be empty");
      }
      if (quote == null) {
        quote = new Quote.LiteralQuoteConfig();
      }
    }

    public FormatConfig withNullString(String newNullString) {
      return new FormatConfig(preset, newNullString, quote);
    }
  }

  public static FormatConfig simpleFormatConfig() {
    return simpleFormatConfig("<null>");
  }

  public static FormatConfig simpleFormatConfig(String nullString) {
    return new FormatConfig(Preset.SIMPLE, nullString, new Quote.LiteralQuoteConfig());
  }

  public static FormatConfig literalFormatConfig() {
    return literalFormatConfig(new Quote.LiteralQuoteConfig(), "NULL");
  }

  public static FormatConfig literalFormatConfig(Quote.LiteralQuoteConfig quote) {
    return literalFormatConfig(quote, "NULL");
  }

  public static FormatConfig literalFormatConfig(Quote.LiteralQuoteConfig quote, String nullString) {
    return new FormatConfig(
        Preset.LITERAL, nullString, Quote.LiteralQuoteConfig.normalize(quote));
  }

  public static FormatConfig spannerCliFormatConfig() {
    return spannerCliFormatConfig("NULL");
  }

  public static FormatConfig spannerCliFormatConfig(String nullString) {
    return new FormatConfig(Preset.SPANNER_CLI, nullString, new Quote.LiteralQuoteConfig());
  }

  private static boolean isComplexType(int code) {
    return code == TypeCode.ARRAY.getValue() || code == TypeCode.STRUCT.getValue();
  }

  private static boolean isScalarType(int code) {
    return code == TypeCode.BOOL.getValue()
        || code == TypeCode.INT64.getValue()
        || code == TypeCode.ENUM.getValue()
        || code == TypeCode.FLOAT32.getValue()
        || code == TypeCode.FLOAT64.getValue()
        || code == TypeCode.STRING.getValue()
        || code == TypeCode.BYTES.getValue()
        || code == TypeCode.PROTO.getValue()
        || code == TypeCode.TIMESTAMP.getValue()
        || code == TypeCode.DATE.getValue()
        || code == TypeCode.NUMERIC.getValue()
        || code == TypeCode.JSON.getValue()
        || code == TypeCode.INTERVAL.getValue()
        || code == TypeCode.UUID.getValue();
  }

  private static void requireStringWire(Object value, int code) {
    if (!"string".equals(valueKind(value))) {
      throw new MalformedWireException(
          formatTypeCodeName(code) + " value kind " + quote(valueKind(value)));
    }
  }

  private static void requireBoolWire(Object value, int code) {
    if (!"bool".equals(valueKind(value))) {
      throw new MalformedWireException(
          formatTypeCodeName(code) + " value kind " + quote(valueKind(value)));
    }
  }

  private static void validateFloatWire(Object value, int code) {
    String kind = valueKind(value);
    if ("number".equals(kind)) {
      return;
    }
    if ("string".equals(kind)) {
      String s = stringValue(value);
      if ("NaN".equals(s) || "Infinity".equals(s) || "-Infinity".equals(s)) {
        return;
      }
      throw new MalformedWireException(
          formatTypeCodeName(code) + " unexpected float string " + quote(s));
    }
    throw new MalformedWireException(
        formatTypeCodeName(code) + " value kind " + quote(kind));
  }

  private static String quote(String s) {
    return "'" + s + "'";
  }

  private static String formatTypeCodeName(int code) {
    return formatTypeCode(code, TypeFormat.UnknownMode.VERBOSE);
  }

  private static double gcvFloat64(Object value) {
    String kind = valueKind(value);
    if ("number".equals(kind)) {
      return numberValue(value);
    }
    if ("string".equals(kind)) {
      String s = stringValue(value);
      return switch (s) {
        case "NaN" -> Double.NaN;
        case "Infinity" -> Double.POSITIVE_INFINITY;
        case "-Infinity" -> Double.NEGATIVE_INFINITY;
        default ->
            throw new MalformedWireException("FLOAT64 unexpected float string " + quote(s));
      };
    }
    throw new MalformedWireException("FLOAT64 value kind " + quote(kind));
  }

  private static double gcvFloat32(Object value) {
    return FloatFmt.narrowFloat32(gcvFloat64(value));
  }

  private static void validateScalarWire(Object typ, Object value) {
    if (typ == null) {
      throw new MalformedWireException("nil type with value kind " + quote(valueKind(value)));
    }
    if (isNullValue(value)) {
      throw new MalformedWireException(
          formatTypeCodeName(typeCode(typ)) + " unexpected null value");
    }
    int code = typeCode(typ);
    if (code == TypeCode.BOOL.getValue()) {
      requireBoolWire(value, code);
    } else if (code == TypeCode.INT64.getValue()
        || code == TypeCode.ENUM.getValue()
        || code == TypeCode.STRING.getValue()
        || code == TypeCode.BYTES.getValue()
        || code == TypeCode.PROTO.getValue()
        || code == TypeCode.TIMESTAMP.getValue()
        || code == TypeCode.DATE.getValue()
        || code == TypeCode.NUMERIC.getValue()
        || code == TypeCode.INTERVAL.getValue()
        || code == TypeCode.UUID.getValue()
        || code == TypeCode.JSON.getValue()) {
      requireStringWire(value, code);
    } else if (code == TypeCode.FLOAT32.getValue() || code == TypeCode.FLOAT64.getValue()) {
      validateFloatWire(value, code);
    } else if (code == TypeCode.TYPE_CODE_UNSPECIFIED.getValue()) {
      throw new UnknownTypeException(String.valueOf(typ));
    } else if (!isScalarType(code)) {
      throw new UnknownTypeException(String.valueOf(typ));
    }
  }

  private static String trimSpannerCliNumericFraction(String s) {
    if (!s.contains(".")) {
      return s;
    }
    s = s.replaceAll("0+$", "");
    return s.replaceAll("\\.$", "");
  }

  private static String numericWireString(Object value) {
    return stringValue(value);
  }

  private static String stringBasedLiteral(
      String typeName, String payload, Quote.LiteralQuoteConfig quote) {
    return typeName + " " + Quote.toStringLiteral(payload, quote);
  }

  private static String formatScalarSimple(Object typ, Object value) {
    validateScalarWire(typ, value);
    int code = typeCode(typ);
    if (code == TypeCode.BOOL.getValue()) {
      return boolValue(value) ? "true" : "false";
    }
    if (code == TypeCode.INT64.getValue()
        || code == TypeCode.ENUM.getValue()
        || code == TypeCode.STRING.getValue()
        || code == TypeCode.TIMESTAMP.getValue()
        || code == TypeCode.DATE.getValue()
        || code == TypeCode.JSON.getValue()
        || code == TypeCode.INTERVAL.getValue()
        || code == TypeCode.UUID.getValue()) {
      return stringValue(value);
    }
    if (code == TypeCode.FLOAT32.getValue()) {
      return FloatFmt.formatGoG(gcvFloat32(value), 32);
    }
    if (code == TypeCode.FLOAT64.getValue()) {
      return FloatFmt.formatGoG(gcvFloat64(value), 64);
    }
    if (code == TypeCode.BYTES.getValue() || code == TypeCode.PROTO.getValue()) {
      return BytesFmt.readableStringFromBase64Wire(stringValue(value));
    }
    if (code == TypeCode.NUMERIC.getValue()) {
      return numericWireString(value);
    }
    throw new UnknownTypeException(String.valueOf(typ));
  }

  private static void validateInt64Wire(String s) {
    try {
      BigInteger bi = new BigInteger(s, 10);
      BigInteger min = BigInteger.valueOf(Long.MIN_VALUE);
      BigInteger max = BigInteger.valueOf(Long.MAX_VALUE);
      if (bi.compareTo(min) < 0 || bi.compareTo(max) > 0) {
        throw new MalformedWireException("INT64 out of range " + quote(s));
      }
    } catch (NumberFormatException e) {
      throw new MalformedWireException("invalid INT64 wire " + quote(s));
    }
  }

  private static String formatScalarLiteral(
      Object typ, Object value, Quote.LiteralQuoteConfig quote) {
    validateScalarWire(typ, value);
    int code = typeCode(typ);
    if (code == TypeCode.BOOL.getValue()) {
      return boolValue(value) ? "true" : "false";
    }
    if (code == TypeCode.INT64.getValue()) {
      String s = stringValue(value);
      validateInt64Wire(s);
      return s;
    }
    if (code == TypeCode.FLOAT32.getValue()) {
      return FloatFmt.float32ToLiteral(gcvFloat32(value), quote);
    }
    if (code == TypeCode.FLOAT64.getValue()) {
      return FloatFmt.float64ToLiteral(gcvFloat64(value), quote);
    }
    if (code == TypeCode.STRING.getValue()) {
      return Quote.toStringLiteral(stringValue(value), quote);
    }
    if (code == TypeCode.BYTES.getValue() || code == TypeCode.PROTO.getValue()) {
      byte[] data = BytesFmt.decodeBase64Wire(stringValue(value));
      return Quote.toBytesLiteral(data, quote);
    }
    if (code == TypeCode.TIMESTAMP.getValue()) {
      return stringBasedLiteral("TIMESTAMP", stringValue(value), quote);
    }
    if (code == TypeCode.DATE.getValue()) {
      return stringBasedLiteral("DATE", stringValue(value), quote);
    }
    if (code == TypeCode.NUMERIC.getValue()) {
      return stringBasedLiteral("NUMERIC", numericWireString(value), quote);
    }
    if (code == TypeCode.JSON.getValue()) {
      return stringBasedLiteral("JSON", stringValue(value), quote);
    }
    if (code == TypeCode.INTERVAL.getValue()) {
      return Quote.sqlCastQuoted(stringValue(value), "INTERVAL", quote);
    }
    if (code == TypeCode.UUID.getValue()) {
      return Quote.sqlCastQuoted(stringValue(value), "UUID", quote);
    }
    throw new UnknownTypeException(String.valueOf(typ));
  }

  private static String formatScalarSpannerCli(Object typ, Object value) {
    validateScalarWire(typ, value);
    int code = typeCode(typ);
    if (code == TypeCode.BOOL.getValue()) {
      return boolValue(value) ? "true" : "false";
    }
    if (code == TypeCode.INT64.getValue()
        || code == TypeCode.ENUM.getValue()
        || code == TypeCode.STRING.getValue()
        || code == TypeCode.BYTES.getValue()
        || code == TypeCode.PROTO.getValue()
        || code == TypeCode.TIMESTAMP.getValue()
        || code == TypeCode.DATE.getValue()
        || code == TypeCode.INTERVAL.getValue()
        || code == TypeCode.UUID.getValue()
        || code == TypeCode.JSON.getValue()) {
      return stringValue(value);
    }
    if (code == TypeCode.FLOAT32.getValue()) {
      return FloatFmt.formatSpannerCliFloat(gcvFloat32(value), 32);
    }
    if (code == TypeCode.FLOAT64.getValue()) {
      return FloatFmt.formatSpannerCliFloat(gcvFloat64(value), 64);
    }
    if (code == TypeCode.NUMERIC.getValue()) {
      return trimSpannerCliNumericFraction(numericWireString(value));
    }
    throw new UnknownTypeException(String.valueOf(typ));
  }

  private static String formatProtoLiteral(
      Object typ, Object value, Quote.LiteralQuoteConfig quote, String nullString) {
    if (typeCode(typ) != TypeCode.PROTO.getValue()) {
      throw new UnknownTypeException(String.valueOf(typ));
    }
    if (isNullValue(value)) {
      return nullString;
    }
    requireStringWire(value, TypeCode.PROTO.getValue());
    byte[] data = BytesFmt.decodeBase64Wire(stringValue(value));
    String fqn = protoTypeFqn(typ);
    if (fqn.isEmpty()) {
      throw new EmptyTypeFQNException("empty type FQN for PROTO");
    }
    return "CAST(" + Quote.toBytesLiteral(data, quote) + " AS `" + fqn + "`)";
  }

  private static String formatEnumLiteral(Object typ, Object value, String nullString) {
    if (typeCode(typ) != TypeCode.ENUM.getValue()) {
      throw new UnknownTypeException(String.valueOf(typ));
    }
    if (isNullValue(value)) {
      return nullString;
    }
    requireStringWire(value, TypeCode.ENUM.getValue());
    String s = stringValue(value);
    try {
      new BigInteger(s, 10);
    } catch (NumberFormatException e) {
      throw new MalformedWireException("failed to parse enum wire payload " + quote(s));
    }
    String fqn = protoTypeFqn(typ);
    if (fqn.isEmpty()) {
      throw new EmptyTypeFQNException("empty type FQN for ENUM");
    }
    return "CAST(" + s + " AS `" + fqn + "`)";
  }

  private static String formatEnumSimple(Object typ, Object value, String nullString) {
    if (isNullValue(value)) {
      return nullString;
    }
    return formatScalarSimple(typ, value);
  }

  private static List<Object> getListValue(Object typ, Object value, int expectedCode) {
    if (!"list".equals(valueKind(value))) {
      throw new UnexpectedComplexValueKindException(
          "unexpected complex value kind for "
              + formatTypeCodeName(expectedCode)
              + ": "
              + quote(valueKind(value)));
    }
    return listValues(value);
  }

  public static String formatValue(Object typ, Object value, FormatConfig config) {
    return formatValue(typ, value, config, true);
  }

  public static String formatValue(
      Object typ, Object value, FormatConfig config, boolean toplevel) {
    if (isNullValue(value)) {
      return config.nullString();
    }

    int code = typeCode(typ);

    if (code == TypeCode.ARRAY.getValue()) {
      List<Object> elems = getListValue(typ, value, code);
      Object elemType = arrayElementType(typ);
      List<String> parts = new ArrayList<>();
      for (Object elem : elems) {
        parts.add(formatValue(elemType, elem, config, false));
      }
      String joined = String.join(", ", parts);
      if (config.preset() == Preset.LITERAL
          && toplevel
          && isComplexType(typeCode(elemType))) {
        return formatTypeVerbose(typ) + "[" + joined + "]";
      }
      return "[" + joined + "]";
    }

    if (code == TypeCode.STRUCT.getValue()) {
      List<Object> fieldVals = getListValue(typ, value, code);
      List<Object> fields = structFields(typ);
      if (fieldVals.size() != fields.size()) {
        throw new MismatchedFieldsException(
            "got " + fieldVals.size() + " values, want " + fields.size());
      }
      if (config.preset() == Preset.SIMPLE) {
        List<String> fieldStrs = new ArrayList<>();
        for (int i = 0; i < fields.size(); i++) {
          String rendered = formatValue(fieldType(fields.get(i)), fieldVals.get(i), config, false);
          String name = fieldName(fields.get(i));
          if (!name.isEmpty()) {
            fieldStrs.add(rendered + " AS " + name);
          } else {
            fieldStrs.add(rendered);
          }
        }
        return "(" + String.join(", ", fieldStrs) + ")";
      }
      List<String> fieldStrs = new ArrayList<>();
      for (int i = 0; i < fields.size(); i++) {
        fieldStrs.add(formatValue(fieldType(fields.get(i)), fieldVals.get(i), config, false));
      }
      String inner = String.join(", ", fieldStrs);
      if (config.preset() == Preset.LITERAL) {
        String prefix = toplevel ? formatTypeVerbose(typ) : "";
        return prefix + "(" + inner + ")";
      }
      if (config.preset() == Preset.SPANNER_CLI) {
        return "[" + inner + "]";
      }
      return "(" + inner + ")";
    }

    if (code == TypeCode.PROTO.getValue()) {
      if (config.preset() == Preset.LITERAL) {
        return formatProtoLiteral(typ, value, config.quote(), config.nullString());
      }
      requireStringWire(value, code);
      if (config.preset() == Preset.SPANNER_CLI) {
        return stringValue(value);
      }
      return BytesFmt.readableStringFromBase64Wire(stringValue(value));
    }

    if (code == TypeCode.ENUM.getValue()) {
      if (config.preset() == Preset.LITERAL) {
        return formatEnumLiteral(typ, value, config.nullString());
      }
      return formatEnumSimple(typ, value, config.nullString());
    }

    if (code == TypeCode.TYPE_CODE_UNSPECIFIED.getValue() || !isScalarType(code)) {
      throw new UnknownTypeException(String.valueOf(typ));
    }

    return switch (config.preset()) {
      case SIMPLE -> formatScalarSimple(typ, value);
      case LITERAL -> formatScalarLiteral(typ, value, config.quote());
      case SPANNER_CLI -> formatScalarSpannerCli(typ, value);
    };
  }

  public static List<String> formatRow(
      List<Object> types, List<Object> values, FormatConfig config) {
    if (types.size() != values.size()) {
      throw new IllegalArgumentException(
          "len(types)=" + types.size() + " != len(values)=" + values.size());
    }
    List<String> out = new ArrayList<>(types.size());
    for (int i = 0; i < types.size(); i++) {
      out.add(formatValue(types.get(i), values.get(i), config, true));
    }
    return out;
  }
}

File: php/src/ValueFormat.php
<?php

declare(strict_types=1);

namespace Apstndb\SpanValue;

enum Preset: int
{
    case SIMPLE = 0;
    case LITERAL = 1;
    case SPANNER_CLI = 2;
}

final readonly class FormatConfig
{
    public function __construct(
        public Preset $preset = Preset::SIMPLE,
        public string $nullString = '<null>',
        public LiteralQuoteConfig $quote = new LiteralQuoteConfig(),
    ) {
        if ($this->nullString === '') {
            throw new EmptyNullStringError('null_string must not be empty');
        }
    }

    public function withNullString(string $nullString): self
    {
        return new self($this->preset, $nullString, $this->quote);
    }
}

final class ValueFormat
{
    public static function simpleFormatConfig(string $nullString = '<null>'): FormatConfig
    {
        return new FormatConfig(Preset::SIMPLE, $nullString);
    }

    public static function literalFormatConfig(
        ?LiteralQuoteConfig $quote = null,
        string $nullString = 'NULL',
    ): FormatConfig {
        $q = Quote::normalizeLiteralQuote($quote ?? new LiteralQuoteConfig());
        return new FormatConfig(Preset::LITERAL, $nullString, $q);
    }

    public static function spannerCliFormatConfig(string $nullString = 'NULL'): FormatConfig
    {
        return new FormatConfig(Preset::SPANNER_CLI, $nullString);
    }

    private static function isComplexType(int $code): bool
    {
        return $code === TypeCode::ARRAY->value || $code === TypeCode::STRUCT->value;
    }

    private static function isScalarType(int $code): bool
    {
        return in_array($code, [
            TypeCode::BOOL->value,
            TypeCode::INT64->value,
            TypeCode::ENUM->value,
            TypeCode::FLOAT32->value,
            TypeCode::FLOAT64->value,
            TypeCode::STRING->value,
            TypeCode::BYTES->value,
            TypeCode::PROTO->value,
            TypeCode::TIMESTAMP->value,
            TypeCode::DATE->value,
            TypeCode::NUMERIC->value,
            TypeCode::JSON->value,
            TypeCode::INTERVAL->value,
            TypeCode::UUID->value,
        ], true);
    }

    private static function requireStringWire(mixed $value, int $code): void
    {
        if (Proto::valueKind($value) !== 'string') {
            throw new MalformedWireError(
                self::formatTypeCodeName($code) . ' value kind ' . var_export(Proto::valueKind($value), true)
            );
        }
    }

    private static function requireBoolWire(mixed $value, int $code): void
    {
        if (Proto::valueKind($value) !== 'bool') {
            throw new MalformedWireError(
                self::formatTypeCodeName($code) . ' value kind ' . var_export(Proto::valueKind($value), true)
            );
        }
    }

    private static function validateFloatWire(mixed $value, int $code): void
    {
        $kind = Proto::valueKind($value);
        if ($kind === 'number') {
            return;
        }
        if ($kind === 'string') {
            $s = Proto::stringValue($value);
            if (in_array($s, ['NaN', 'Infinity', '-Infinity'], true)) {
                return;
            }
            throw new MalformedWireError(
                self::formatTypeCodeName($code) . ' unexpected float string ' . var_export($s, true)
            );
        }
        throw new MalformedWireError(
            self::formatTypeCodeName($code) . ' value kind ' . var_export($kind, true)
        );
    }

    public static function formatTypeCodeName(int $code): string
    {
        return TypeFormat::formatTypeCode($code);
    }

    private static function gcvFloat64(mixed $value): float
    {
        $kind = Proto::valueKind($value);
        if ($kind === 'number') {
            return Proto::numberValue($value);
        }
        if ($kind === 'string') {
            $s = Proto::stringValue($value);
            return match ($s) {
                'NaN' => NAN,
                'Infinity' => INF,
                '-Infinity' => -INF,
                default => throw new MalformedWireError("FLOAT64 unexpected float string {$s}"),
            };
        }
        throw new MalformedWireError('FLOAT64 value kind ' . var_export($kind, true));
    }

    private static function gcvFloat32(mixed $value): float
    {
        return FloatFmt::narrowFloat32(self::gcvFloat64($value));
    }

    private static function validateScalarWire(mixed $typ, mixed $value): void
    {
        if ($typ === null) {
            throw new MalformedWireError('nil type with value kind ' . var_export(Proto::valueKind($value), true));
        }
        if (Proto::isNullValue($value)) {
            throw new MalformedWireError(
                self::formatTypeCodeName(Proto::typeCode($typ)) . ' unexpected null value'
            );
        }
        $code = Proto::typeCode($typ);
        if ($code === TypeCode::BOOL->value) {
            self::requireBoolWire($value, $code);
        } elseif (in_array($code, [
            TypeCode::INT64->value,
            TypeCode::ENUM->value,
            TypeCode::STRING->value,
            TypeCode::BYTES->value,
            TypeCode::PROTO->value,
            TypeCode::TIMESTAMP->value,
            TypeCode::DATE->value,
            TypeCode::NUMERIC->value,
            TypeCode::INTERVAL->value,
            TypeCode::UUID->value,
            TypeCode::JSON->value,
        ], true)) {
            self::requireStringWire($value, $code);
        } elseif ($code === TypeCode::FLOAT32->value || $code === TypeCode::FLOAT64->value) {
            self::validateFloatWire($value, $code);
        } elseif ($code === TypeCode::TYPE_CODE_UNSPECIFIED->value) {
            throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
        } elseif (!self::isScalarType($code)) {
            throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
        }
    }

    private static function isValidInt64Wire(string $s): bool
    {
        if (!preg_match('/^-?\d+$/', $s)) {
            return false;
        }
        $min = '-9223372036854775808';
        $max = '9223372036854775807';
        if (str_starts_with($s, '-')) {
            return strlen($s) < strlen($min) || (strlen($s) === strlen($min) && $s >= $min);
        }
        return strlen($s) < strlen($max) || (strlen($s) === strlen($max) && $s <= $max);
    }

    private static function trimSpannerCliNumericFraction(string $s): string
    {
        if (!str_contains($s, '.')) {
            return $s;
        }
        $s = rtrim($s, '0');
        return rtrim($s, '.');
    }

    private static function numericWireString(mixed $value): string
    {
        return Proto::stringValue($value);
    }

    private static function stringBasedLiteral(string $typeName, string $payload, LiteralQuoteConfig $quote): string
    {
        return "{$typeName} " . Quote::toStringLiteral($payload, $quote);
    }

    private static function formatScalarSimple(mixed $typ, mixed $value): string
    {
        self::validateScalarWire($typ, $value);
        $code = Proto::typeCode($typ);
        if ($code === TypeCode::BOOL->value) {
            return Proto::boolValue($value) ? 'true' : 'false';
        }
        if (in_array($code, [
            TypeCode::INT64->value,
            TypeCode::ENUM->value,
            TypeCode::STRING->value,
            TypeCode::TIMESTAMP->value,
            TypeCode::DATE->value,
            TypeCode::JSON->value,
            TypeCode::INTERVAL->value,
            TypeCode::UUID->value,
        ], true)) {
            return Proto::stringValue($value);
        }
        if ($code === TypeCode::FLOAT32->value) {
            return FloatFmt::formatGoG(self::gcvFloat32($value), 32);
        }
        if ($code === TypeCode::FLOAT64->value) {
            return FloatFmt::formatGoG(self::gcvFloat64($value), 64);
        }
        if ($code === TypeCode::BYTES->value || $code === TypeCode::PROTO->value) {
            return BytesFmt::readableStringFromBase64Wire(Proto::stringValue($value));
        }
        if ($code === TypeCode::NUMERIC->value) {
            return self::numericWireString($value);
        }
        throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
    }

    private static function formatScalarLiteral(mixed $typ, mixed $value, LiteralQuoteConfig $quote): string
    {
        self::validateScalarWire($typ, $value);
        $code = Proto::typeCode($typ);
        if ($code === TypeCode::BOOL->value) {
            return Proto::boolValue($value) ? 'true' : 'false';
        }
        if ($code === TypeCode::INT64->value) {
            $s = Proto::stringValue($value);
            if (!self::isValidInt64Wire($s)) {
                throw new MalformedWireError("invalid INT64 wire {$s}");
            }
            return $s;
        }
        if ($code === TypeCode::FLOAT32->value) {
            return FloatFmt::float32ToLiteral(self::gcvFloat32($value), $quote);
        }
        if ($code === TypeCode::FLOAT64->value) {
            return FloatFmt::float64ToLiteral(self::gcvFloat64($value), $quote);
        }
        if ($code === TypeCode::STRING->value) {
            return Quote::toStringLiteral(Proto::stringValue($value), $quote);
        }
        if ($code === TypeCode::BYTES->value || $code === TypeCode::PROTO->value) {
            $data = BytesFmt::decodeBase64Wire(Proto::stringValue($value));
            return Quote::toBytesLiteral($data, $quote);
        }
        if ($code === TypeCode::TIMESTAMP->value) {
            return self::stringBasedLiteral('TIMESTAMP', Proto::stringValue($value), $quote);
        }
        if ($code === TypeCode::DATE->value) {
            return self::stringBasedLiteral('DATE', Proto::stringValue($value), $quote);
        }
        if ($code === TypeCode::NUMERIC->value) {
            return self::stringBasedLiteral('NUMERIC', self::numericWireString($value), $quote);
        }
        if ($code === TypeCode::JSON->value) {
            return self::stringBasedLiteral('JSON', Proto::stringValue($value), $quote);
        }
        if ($code === TypeCode::INTERVAL->value) {
            return Quote::sqlCastQuoted(Proto::stringValue($value), 'INTERVAL', $quote);
        }
        if ($code === TypeCode::UUID->value) {
            return Quote::sqlCastQuoted(Proto::stringValue($value), 'UUID', $quote);
        }
        throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
    }

    private static function formatScalarSpannerCli(mixed $typ, mixed $value): string
    {
        self::validateScalarWire($typ, $value);
        $code = Proto::typeCode($typ);
        if ($code === TypeCode::BOOL->value) {
            return Proto::boolValue($value) ? 'true' : 'false';
        }
        if (in_array($code, [
            TypeCode::INT64->value,
            TypeCode::ENUM->value,
            TypeCode::STRING->value,
            TypeCode::BYTES->value,
            TypeCode::PROTO->value,
            TypeCode::TIMESTAMP->value,
            TypeCode::DATE->value,
            TypeCode::INTERVAL->value,
            TypeCode::UUID->value,
            TypeCode::JSON->value,
        ], true)) {
            return Proto::stringValue($value);
        }
        if ($code === TypeCode::FLOAT32->value) {
            return FloatFmt::formatSpannerCliFloat(self::gcvFloat32($value), 32);
        }
        if ($code === TypeCode::FLOAT64->value) {
            return FloatFmt::formatSpannerCliFloat(self::gcvFloat64($value), 64);
        }
        if ($code === TypeCode::NUMERIC->value) {
            return self::trimSpannerCliNumericFraction(self::numericWireString($value));
        }
        throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
    }

    private static function formatProtoLiteral(
        mixed $typ,
        mixed $value,
        LiteralQuoteConfig $quote,
        string $nullString,
    ): string {
        if (Proto::typeCode($typ) !== TypeCode::PROTO->value) {
            throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
        }
        if (Proto::isNullValue($value)) {
            return $nullString;
        }
        self::requireStringWire($value, TypeCode::PROTO->value);
        $data = BytesFmt::decodeBase64Wire(Proto::stringValue($value));
        $fqn = Proto::protoTypeFqn($typ);
        if ($fqn === '') {
            throw new EmptyTypeFQNError('empty type FQN for PROTO');
        }
        return 'CAST(' . Quote::toBytesLiteral($data, $quote) . " AS `{$fqn}`)";
    }

    private static function formatEnumLiteral(mixed $typ, mixed $value, string $nullString): string
    {
        if (Proto::typeCode($typ) !== TypeCode::ENUM->value) {
            throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
        }
        if (Proto::isNullValue($value)) {
            return $nullString;
        }
        self::requireStringWire($value, TypeCode::ENUM->value);
        $s = Proto::stringValue($value);
        if (!preg_match('/^-?\d+$/', $s)) {
            throw new MalformedWireError("failed to parse enum wire payload {$s}");
        }
        $fqn = Proto::protoTypeFqn($typ);
        if ($fqn === '') {
            throw new EmptyTypeFQNError('empty type FQN for ENUM');
        }
        return "CAST({$s} AS `{$fqn}`)";
    }

    private static function formatEnumSimple(mixed $typ, mixed $value, string $nullString): string
    {
        if (Proto::isNullValue($value)) {
            return $nullString;
        }
        return self::formatScalarSimple($typ, $value);
    }

    /** @return list<mixed> */
    private static function getListValue(mixed $typ, mixed $value, int $expectedCode): array
    {
        if (Proto::valueKind($value) !== 'list') {
            throw new UnexpectedComplexValueKindError(
                'unexpected complex value kind for '
                . self::formatTypeCodeName($expectedCode)
                . ': ' . var_export(Proto::valueKind($value), true)
            );
        }
        return Proto::listValues($value);
    }

    public static function formatValue(
        mixed $typ,
        mixed $value,
        FormatConfig $config,
        bool $toplevel = true,
    ): string {
        if (Proto::isNullValue($value)) {
            return $config->nullString;
        }

        $code = Proto::typeCode($typ);

        if ($code === TypeCode::ARRAY->value) {
            $elems = self::getListValue($typ, $value, $code);
            $elemType = Proto::arrayElementType($typ);
            $parts = [];
            foreach ($elems as $elem) {
                $parts[] = self::formatValue($elemType, $elem, $config, false);
            }
            $joined = implode(', ', $parts);
            if (
                $config->preset === Preset::LITERAL
                && $toplevel
                && self::isComplexType(Proto::typeCode($elemType))
            ) {
                return TypeFormat::formatTypeVerbose($typ) . "[{$joined}]";
            }
            return "[{$joined}]";
        }

        if ($code === TypeCode::STRUCT->value) {
            $fieldVals = self::getListValue($typ, $value, $code);
            $fields = Proto::structFields($typ);
            if (count($fieldVals) !== count($fields)) {
                throw new MismatchedFieldsError(
                    'got ' . count($fieldVals) . ' values, want ' . count($fields)
                );
            }
            if ($config->preset === Preset::SIMPLE) {
                $fieldStrs = [];
                foreach ($fields as $i => $field) {
                    $rendered = self::formatValue(Proto::fieldType($field), $fieldVals[$i], $config, false);
                    $name = Proto::fieldName($field);
                    if ($name !== '') {
                        $fieldStrs[] = "{$rendered} AS {$name}";
                    } else {
                        $fieldStrs[] = $rendered;
                    }
                }
                return '(' . implode(', ', $fieldStrs) . ')';
            }
            $fieldStrs = [];
            foreach ($fields as $i => $field) {
                $fieldStrs[] = self::formatValue(Proto::fieldType($field), $fieldVals[$i], $config, false);
            }
            $inner = implode(', ', $fieldStrs);
            if ($config->preset === Preset::LITERAL) {
                $prefix = $toplevel ? TypeFormat::formatTypeVerbose($typ) : '';
                return "{$prefix}({$inner})";
            }
            if ($config->preset === Preset::SPANNER_CLI) {
                return "[{$inner}]";
            }
            return "({$inner})";
        }

        if ($code === TypeCode::PROTO->value) {
            if ($config->preset === Preset::LITERAL) {
                return self::formatProtoLiteral($typ, $value, $config->quote, $config->nullString);
            }
            self::requireStringWire($value, $code);
            if ($config->preset === Preset::SPANNER_CLI) {
                return Proto::stringValue($value);
            }
            return BytesFmt::readableStringFromBase64Wire(Proto::stringValue($value));
        }

        if ($code === TypeCode::ENUM->value) {
            if ($config->preset === Preset::LITERAL) {
                return self::formatEnumLiteral($typ, $value, $config->nullString);
            }
            return self::formatEnumSimple($typ, $value, $config->nullString);
        }

        if ($code === TypeCode::TYPE_CODE_UNSPECIFIED->value || !self::isScalarType($code)) {
            throw new UnknownTypeError(json_encode($typ) ?: (string) $typ);
        }

        return match ($config->preset) {
            Preset::SIMPLE => self::formatScalarSimple($typ, $value),
            Preset::LITERAL => self::formatScalarLiteral($typ, $value, $config->quote),
            Preset::SPANNER_CLI => self::formatScalarSpannerCli($typ, $value),
        };
    }

    /** @param list<mixed> $types @param list<mixed> $values @return list<string> */
    public static function formatRow(array $types, array $values, FormatConfig $config): array
    {
        if (count($types) !== count($values)) {
            throw new \InvalidArgumentException(
                'len(types)=' . count($types) . ' != len(values)=' . count($values)
            );
        }
        $result = [];
        foreach ($types as $i => $type) {
            $result[] = self::formatValue($type, $values[$i], $config, true);
        }
        return $result;
    }
}
